From 7c209f25f2c14e4d9cf6abb1b5f3035b94341870 Mon Sep 17 00:00:00 2001
From: Ben Greear <greearb@candelatech.com>
Date: Fri, 21 May 2021 15:56:33 -0700
Subject: [PATCH 07/15] intel-oot:  Add backport compat code to iwl-config.h

This adds enough of the backport related compat code to
get this to (mostly) compile.

Signed-off-by: Ben Greear <greearb@candelatech.com>
---
 .../wireless/intel-oot/iwlwifi/iwl-config.h   | 79 +++++++++++++++++++
 1 file changed, 79 insertions(+)

diff --git a/drivers/net/wireless/intel-oot/iwlwifi/iwl-config.h b/drivers/net/wireless/intel-oot/iwlwifi/iwl-config.h
index b604a8e6ff76e..41e5c2b94da2e 100644
--- a/drivers/net/wireless/intel-oot/iwlwifi/iwl-config.h
+++ b/drivers/net/wireless/intel-oot/iwlwifi/iwl-config.h
@@ -12,6 +12,85 @@
 #include <linux/nl80211.h>
 #include "iwl-csr.h"

+/* Start of backport hackings */
+#include <linux/version.h>
+#define LINUX_VERSION_IS_LESS(x1,x2,x3) (LINUX_VERSION_CODE < KERNEL_VERSION(x1,x2,x3))
+#define LINUX_VERSION_IS_GEQ(x1,x2,x3)  (LINUX_VERSION_CODE >= KERNEL_VERSION(x1,x2,x3))
+#define LINUX_VERSION_IN_RANGE(x1,x2,x3, y1,y2,y3) \
+        (LINUX_VERSION_IS_GEQ(x1,x2,x3) && LINUX_VERSION_IS_LESS(y1,y2,y3))
+
+#if LINUX_VERSION_IS_LESS(3,7,0)
+#define genl_info_snd_portid(__genl_info) (__genl_info->snd_pid)
+#else
+#define genl_info_snd_portid(__genl_info) (__genl_info->snd_portid)
+#endif
+
+#if LINUX_VERSION_IS_LESS(3,13,0)
+#define __genl_const
+#else /* < 3.13 */
+#define __genl_const const
+#endif /* < 3.13 */
+
+#ifndef GENLMSG_DEFAULT_SIZE
+#define GENLMSG_DEFAULT_SIZE (NLMSG_DEFAULT_SIZE - GENL_HDRLEN)
+#endif
+
+#if LINUX_VERSION_IS_LESS(3,1,0)
+#define genl_dump_check_consistent(cb, user_hdr) do { } while (0)
+#else
+#define netlink_notify_portid(__notify) (__notify->portid)
+#define NETLINK_CB_PORTID(__skb) NETLINK_CB(__skb).portid
+#endif
+
+#if LINUX_VERSION_IS_LESS(4,10,0)
+#define __genl_ro_after_init
+#else
+#define __genl_ro_after_init __ro_after_init
+#endif
+
+#if LINUX_VERSION_IS_LESS(3,7,0)
+#define netlink_notify_portid(__notify) (__notify->pid)
+#define NETLINK_CB_PORTID(__skb) NETLINK_CB(__skb).pid
+#endif
+
+#define BACKPORTS_GIT_TRACKED "ct-5.11"
+#define BACKPORTS_BUILD_TSTAMP "NA"
+
+#if LINUX_VERSION_IS_LESS(3,17,0)
+#define ktime_get_ns LINUX_BACKPORT(ktime_get_ns)
+extern ktime_t ktime_get(void);
+#define ktime_get_ns LINUX_BACKPORT(ktime_get_ns)
+static inline u64 ktime_get_ns(void)
+{
+        return ktime_to_ns(ktime_get());
+}
+
+extern ktime_t ktime_get_boottime(void);
+#define ktime_get_boot_ns LINUX_BACKPORT(ktime_get_boot_ns)
+static inline u64 ktime_get_boot_ns(void)
+{
+        return ktime_to_ns(ktime_get_boottime());
+}
+#endif /* < 3.17 */
+
+#if LINUX_VERSION_IS_GEQ(5,3,0)
+/*
+ * In v5.3, this function was renamed, so rename it here for v5.3+.
+ * When we merge v5.3 back from upstream, the opposite should be done
+ * (i.e. we will have _boottime_ and need to rename to _boot_ in <
+ * v5.3 instead).
+*/
+#define ktime_get_boot_ns ktime_get_boottime_ns
+#endif /* > 5.3.0 */
+
+#if LINUX_VERSION_IS_LESS(4,18,0)
+extern time64_t ktime_get_boottime_seconds(void);
+#endif /* < 4.18 */
+
+#define CONFIG_OOT_IWL_TIMEOUT_FACTOR 1
+
+/* End of backport hackings */
+
 enum iwl_device_family {
	IWL_DEVICE_FAMILY_UNDEFINED,
	IWL_DEVICE_FAMILY_1000,
--
2.26.3
